torchvision/prototype/datasets/utils/_dataset.py

```diff
@@ -0,0 +1,218 @@
+import abc
+import io
+import os
+import pathlib
+import textwrap
+from collections import Mapping
+from typing import (
+    Any,
+    Callable,
+    Dict,
+    List,
+    Optional,
+    Sequence,
+    Union,
+    NoReturn,
+    Iterable,
+    Tuple,
+)
+
+import torch
+from torch.utils.data import IterDataPipe
+
+from torchvision.prototype.datasets.utils._internal import (
+    add_suggestion,
+    sequence_to_str,
+)
+from ._resource import OnlineResource
+
+
+def make_repr(name: str, items: Iterable[Tuple[str, Any]]):
+    def to_str(sep: str) -> str:
+        return sep.join([f"{key}={value}" for key, value in items])
+
+    prefix = f"{name}("
+    postfix = ")"
+    body = to_str(", ")
+
+    line_length = int(os.environ.get("COLUMNS", 80))
+    body_too_long = (len(prefix) + len(body) + len(postfix)) > line_length
+    multiline_body = len(str(body).splitlines()) > 1
+    if not (body_too_long or multiline_body):
+        return prefix + body + postfix
+
+    body = textwrap.indent(to_str(",\n"), " " * 2)
+    return f"{prefix}\n{body}\n{postfix}"
+
+
+class DatasetConfig(Mapping):
```

Comment:
This class looks a lot like overkill, but it is actually not:

1. Since `lambda`'s are "not welcome" in datapipes that take a callable (a warning will be thrown), we often need to use `functools.partial(..., config=config)`. To do that `config` has to be hashable. Hence, we implement the `__hash__` and `__eq__` methods and fail in case `__getitem__` or `__delitem__` are invoked.
2. For convenience it is much nicer to use a `DatasetConfig` as namespace as a dictionary, so we additionally implement `__getattr__`, `__setattr__`, and `__delattr__`.